File: android-client/lib/services/game_service.dart
import 'dart:async';
import 'dart:math';
import 'dart:typed_data';

import 'package:android_client/classes/Vec2.dart';
import 'package:android_client/classes/difference_try.dart';
import 'package:android_client/classes/differences_hashmap.dart';
import 'package:android_client/classes/end_Game.dart';
import 'package:android_client/classes/game_constants.dart';
import 'package:android_client/classes/game_data.dart';
import 'package:android_client/classes/game_room.dart';
import 'package:android_client/classes/replay.dart';
import 'package:android_client/classes/user_game.dart';
import 'package:android_client/services/remote_games_manager_service.dart';
import 'package:android_client/services/socket_service.dart';
import 'package:android_client/services/video_replay_service.dart';
import 'package:http/http.dart' as http;

class HintData {
  String imageData;
  bool left;
  HintData({required this.imageData, required this.left});
}

class GameService {
  StreamController<bool> gameExistsController = StreamController<bool>();
  Stream<bool> get gameExists$ => gameExistsController.stream;

  StreamController<int> timePositionController = StreamController<int>();
  Stream<int> get timePosition$ => timePositionController.stream;

  StreamController<DifferenceTry> serverValidateResponseController =
      StreamController<DifferenceTry>();
  Stream<DifferenceTry> get serverValidateResponse$ =>
      serverValidateResponseController.stream;

  StreamController<int> totalDifferencesFoundController =
      StreamController<int>();
  Stream<int> get totalDifferencesFound$ =>
      totalDifferencesFoundController.stream;

  StreamController<int> userDifferencesFoundController =
      StreamController<int>();
  Stream<int> get userDifferencesFound$ =>
      userDifferencesFoundController.stream;

  StreamController<int> timerController = StreamController<int>();
  Stream<int> get timer$ => timerController.stream;

  StreamController<EndGame> gameFinishedController =
      StreamController<EndGame>();
  Stream<EndGame> get gameFinished$ => gameFinishedController.stream;

  StreamController<GameRoom> gameRoomController = StreamController<GameRoom>();
  Stream<GameRoom> get gameRoom$ => gameRoomController.stream;

  StreamController<GameData> gameDataController = StreamController<GameData>();
  Stream<GameData> get gameData$ => gameDataController.stream;

  StreamController<HintData> hintController = StreamController<HintData>();
  Stream<HintData> get hint$ => hintController.stream;

  StreamController<bool> gameDeletedController = StreamController<bool>();
  Stream<bool> get gameDeleted$ => gameDeletedController.stream;

  StreamController<String> abandonedController = StreamController<String>();
  Stream<String> get abandoned$ => abandonedController.stream;

  StreamController<bool> cheatModeResponseController = StreamController<bool>();
  Stream<bool> get cheatModeResponse$ => cheatModeResponseController.stream;

  late GameRoom gameRoom;
  late List<GameData> slides;
  late Map<String, List<DifferencesHashMap>> differencesHashMap =
      <String, List<DifferencesHashMap>>{};
  late GameData gameData;
  late String username;
  late String gameMode;
  Map<String, List<Uint8List>> imagesData = <String, List<Uint8List>>{};
  late GameConstants gameConstants;
  bool canSendValidate = true;
  final VideoReplayService videoReplayService = VideoReplayService();

  final SocketService socketService = SocketService();
  final RemoteGamesManagerService _remoteGamesManagerService =
      RemoteGamesManagerService();
  static final GameService _instance = GameService._internal();

  factory GameService() {
    return _instance;
  }

  GameService._internal();

  void getConstant() {
    gameConstants = gameRoom.gameConstants!;
  }

  bool isLimitedTimeMode() {
    return gameMode == 'mode Temps Limité';
  }

  Future<void> startGame(GameRoom gameRoom, String username) async {
    this.gameRoom = gameRoom;
    this.username = username;
    gameMode = gameRoom.gameMode;
    gameData = await getGameData(gameRoom.userGame.gameName!);
    if (gameRoom.userGame.chosenDifference != -1) {
      gameData.differenceMatrix = gameData
          .differenceHashMap?[gameRoom.userGame.chosenDifference]
          .differenceMatrix;
    }
    gameRoomController.add(gameRoom);
    if (gameRoom.userGame.creator == username) {
      socketService.send('start', gameRoom.roomId);
    }
    if (gameMode == 'mode Temps Limité') {
      slides.removeWhere((GameData slide) => slide.name == gameData.name);
      gameDeletedSocket();
    }
    // Handle Message
    handleSocket();
  }

  Future<void> observeGame(GameRoom gameRoom, String username) async {
    this.gameRoom = gameRoom;
    this.username = username;
    gameMode = gameRoom.gameMode;
    gameData = await getGameData(gameRoom.userGame.gameName!);
    gameRoomController.add(gameRoom);
    if (gameRoom.userGame.chosenDifference != -1) {
      if (gameData.differenceHashMap == null) {
        _remoteGamesManagerService.getGame(gameData.name).then((GameData game) {
          gameData.differenceHashMap = game.differenceHashMap;
          gameData.differenceMatrix = gameData
              .differenceHashMap![gameRoom.userGame.chosenDifference]
              .differenceMatrix;
        });
      } else {
        gameData.differenceMatrix = gameData
            .differenceHashMap![gameRoom.userGame.chosenDifference]
            .differenceMatrix;
      }
    }
    if (gameMode == 'mode Temps Limité') {
      slides.removeWhere((GameData slide) => slide.name == gameData.name);
      gameDeletedSocket();
    }
    // Handle Message
    handleSocket();
    socketService.send('observingGame', gameRoom.toJson());
  }

  void sendHint(String imageData, String? receiver, bool left) {
    socketService.send('hint', <String, Object?>{
      "imageData": imageData,
      "sender": username,
      "receiver": receiver,
      "roomId": gameRoom.roomId,
      "left": left
    });
  }

  void gameDeletedSocket() {
    socketService.on('gameCanceled', (data) {
      slides.removeWhere((GameData slide) => slide.name == data);
    });
  }

  void sendServerDifference(List<List<int>> differenceMatrix, String username) {
    if (this.username == username) {
      socketService.send('sendDifference', <String, Object>{
        "differenceMatrix": differenceMatrix,
        "roomId": gameRoom.roomId
      });
    }
  }

  void turnOffWaitingSocket() {
    socketService.off('gameInfo');
    socketService.off('gameCreated');
    socketService.off('playerAccepted');
    socketService.off('playerRejected');
    socketService.off('gameCanceled');
  }

  void turnOffGameSocket() {
    socketService.off('validated');
    socketService.off('GameFinished');
    socketService.off('abandoned');
    socketService.off('timer');
    socketService.off('cheatMode');
    socketService.off('gameFinished');
    socketService.off('nextGame');
    socketService.off('observingGame');
    socketService.off('hint');
  }

  void getAllGames() async {
    _remoteGamesManagerService.getAllRemoteGames().then((List<GameData> games) {
      games.forEach((GameData game) => _remoteGamesManagerService
              .getGame(game.name)
              .then((GameData gameData) async {
            game.differenceHashMap = gameData.differenceHashMap;
            game.differenceMatrix = gameData.differenceMatrix;
            final Uint8List imageFuture1 = await getImageData(game.image1url);
            final Uint8List imageFuture2 = await getImageData(game.image2url);
            imagesData.putIfAbsent(
                game.name, () => <Uint8List>[imageFuture1, imageFuture2]);
            slides = games;
          }));
    });
  }

  Future<Uint8List> getImageData(String imageUrl) async {
    final http.Response imageResponse = await http
        .get(Uri.parse(_remoteGamesManagerService.getReplacedUrl(imageUrl)));
    return imageResponse.bodyBytes;
  }

  void abortGame() {
    if (gameRoom.userGame.creator == username) {
      socketService.send('abortGameCreation', gameRoom.roomId);
    } else {
      socketService.send('leaveGame',
          <String, String>{"roomId": gameRoom.roomId, "username": username});
    }
  }

  void sendServerValidate(Vec2 differencePos) {
    if (canSendValidate) {
      canSendValidate = false;
      socketService.send('validate', <String, Object>{
        "differencePos": differencePos.toJson(),
        "roomId": gameRoom.roomId,
        "username": username,
        "validated": true
      });
    }
  }

  List<List<int>>? findDifference(Vec2 differencePos) {
    for (DifferencesHashMap differenceHashMap in gameData.differenceHashMap!) {
      print(gameData.name);
      if (differenceHashMap.differenceMatrix[differencePos.y.toInt()]
              [differencePos.x.toInt()] !=
          -1) {
        return differenceHashMap.differenceMatrix;
      }
    }
    return null;
  }

  void changeTime(int number) {
    socketService.send('changeTime',
        <String, Object>{"number": number, "roomId": gameRoom.roomId});
  }

  void reset() {
    // TODO: choose what to reset
  }

  void abandonGame() {
    if (gameRoom.userGame.currentPlayers
        .where((CurrentPlayer user) => user.username == username)
        .any((CurrentPlayer user) => true)) {
      socketService.send('abandoned',
          <String, String>{"roomId": gameRoom.roomId, "username": username});
    } else {
      socketService.send('observerLeaveGame',
          <String, String>{"roomId": gameRoom.roomId, "username": username});
    }
  }

  void loadNextGame() {
    if (slides.isNotEmpty && gameRoom.userGame.creator == username) {
      GameData game = slides[generateRandomNumber(slides.length)];
      gameRoom.userGame.gameName = game.name;
      gameRoom.userGame.chosenDifference =
          generateRandomNumber(game.nbDifference);
      socketService.send('nextGame', gameRoom.toJson());
    } else if (slides.isEmpty) {
      socketService.send('endGame', gameRoom.roomId);
    }
  }

  void isCheatModeOn() {
    if (gameRoom.gameConstants?.cheatMode == false) {
      cheatModeResponseController.add(false);
    }
    socketService.send('cheatMode', gameRoom.roomId);
  }

  Future<GameData> getGameData(String gameName) async {
    if (slides.isEmpty) {
      GameData g = await _remoteGamesManagerService.getGame(gameName);
      slides.add(g);
      return g;
    }
    try {
      return slides.firstWhere((GameData game) => game.name == gameName);
    } catch (e) {
      GameData g = await _remoteGamesManagerService.getGame(gameName);
      slides.add(g);
      return g;
    }
  }

  void sendMatrixToServer(List<List<int>> differenceMatrix, String username) {
    if (this.username == username) {
      socketService.send('sendDifference', <String, Object>{
        "differenceMatrix": differenceMatrix,
        "roomId": gameRoom.roomId
      });
    }
  }

  void handleSocket() {
    socketService.on('started', (data) {
      gameRoomController.add(gameRoom);
      if (gameRoom.userGame.gameName != gameData.name) {
        getGameData(gameRoom.userGame.gameName!).then((GameData game) {
          gameData = game;
          gameData.differenceMatrix = gameData
              .differenceHashMap?[gameRoom.userGame.chosenDifference]
              .differenceMatrix;
          gameDataController.add(gameData);
          videoReplayService.setGameData(gameData);
        });
      }
      videoReplayService.setGameData(gameData);
      if (gameMode == 'mode Temps Limité') return;
      videoReplayService.recordEvent(ReplayEvent(
          action: 'gameStart',
          imageData1: '',
          imageData2: '',
          timestamp: DateTime.now().millisecondsSinceEpoch));
    });

    socketService.on('nextGame', (data) {
      if (data == null) {
        print('No more games');
        return;
      }
      gameRoom = GameRoom.fromJson(data);
      final int slideIndex = slides.indexWhere(
          (GameData game) => game.name == gameRoom.userGame.gameName);
      if (slideIndex != -1) {
        final GameData slide = slides.removeAt(slideIndex);
        gameData = slide;
        gameData.differenceMatrix = gameData
            .differenceHashMap?[gameRoom.userGame.chosenDifference]
            .differenceMatrix;
        gameRoomController.add(gameRoom);
        gameDataController.add(gameData);
      } else {
        print('No more games');
      }
    });

    socketService.on('hint', (data) {
      if (data['receiver'] == username ||
          (data['receiver'] == null &&
              !gameRoom.userGame.observers!
                  .where((Observer player) => player.username == username)
                  .any((Observer player) => true)) ||
          data['sender'] == username) {
        hintController.add(HintData(
            imageData: data['imageData'], left: data['left'] ?? false));
      }
    });

    socketService.on('observingGame', (data) {
      final GameRoom gameRoomSocket = GameRoom.fromJson(data);
      gameRoom.userGame.observers = gameRoomSocket.userGame.observers;
      gameRoomController.add(gameRoom);
    });

    socketService.on('validated', (data) {
      DifferenceTry differenceTry = DifferenceTry.fromJson(data);
      if (differenceTry.validated) {
        gameRoom.userGame.nbDifferenceFound += 1;
        totalDifferencesFoundController
            .add(gameRoom.userGame.nbDifferenceFound);

        gameRoom.userGame.differenceFoundByPlayers =
            differenceTry.playerDifferences;
      }
      serverValidateResponseController.add(differenceTry);
    });

    socketService.on('cheatMode', (data) {
      cheatModeResponseController.add(data);
    });

    socketService.on('GameFinished', (data) {
      EndGame endGame = EndGame.fromJson(data);
      gameFinishedController.add(endGame);
      if (gameMode == 'mode Temps Limité') return;
      videoReplayService.recordEvent(ReplayEvent(
          action: 'gameEnd',
          imageData1: '',
          imageData2: '',
          timestamp: DateTime.now().millisecondsSinceEpoch));
    });

    socketService.on('abandoned', (data) {
      gameRoom = GameRoom.fromJson(data['gameRoom']);
      // WE DON'T NEED THIS BUT I WILL LEAVE IT HERE TO MAKE SURE
      gameRoomController.add(gameRoom);
      abandonedController.add(data['username']);
      // if (gameMode == 'mode Temps Limité') {
      //   limitedTimeGameAbandoned(gameRoom);
      // }
    });

    socketService.on('timer', (data) {
      gameRoom.userGame.timer = data.toInt();
      timerController.add(data.toInt());
      canSendValidate = true;
    });
  }

  // void limitedTimeGameAbandoned(GameRoom gameRoom) {
  //   this.gameRoom = gameRoom;
  //   gameRoomController.add(gameRoom);
  // }

  int generateRandomNumber(int nbOfDifferences) {
    Random random = Random();
    return random.nextInt(nbOfDifferences);
  }

  void closeStreams() {
    turnOffGameSocket();
    gameExistsController.close();
    timePositionController.close();
    serverValidateResponseController.close();
    totalDifferencesFoundController.close();
    userDifferencesFoundController.close();
    timerController.close();
    gameFinishedController.close();
    gameRoomController.close();
    gameDeletedController.close();
    abandonedController.close();
    cheatModeResponseController.close();
    gameDataController.close();
    hintController.close();
    gameExistsController = StreamController<bool>();
    timePositionController = StreamController<int>();
    serverValidateResponseController = StreamController<DifferenceTry>();
    totalDifferencesFoundController = StreamController<int>();
    userDifferencesFoundController = StreamController<int>();
    timerController = StreamController<int>();
    gameFinishedController = StreamController<EndGame>();
    gameRoomController = StreamController<GameRoom>();
    gameDeletedController = StreamController<bool>();
    abandonedController = StreamController<String>();
    cheatModeResponseController = StreamController<bool>();
    gameDataController = StreamController<GameData>();
    hintController = StreamController<HintData>();
  }
}

File: windows-client/src/app/services/game/game.service.ts
import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { MessageDialogComponent } from '@app/components/message-dialog/message-dialog.component';
import { DifferenceTry } from '@app/interfaces/difference-try';
import { EndGame } from '@app/interfaces/game';
import { Vec2 } from '@app/interfaces/vec2';
import { ChatService } from '@app/services/chat/chat.service';
import { CommunicationHttpService } from '@app/services/communication-http/communication-http.service';
import { CommunicationSocketService } from '@app/services/communication-socket/communication-socket.service';
import { VideoReplayService } from '@app/services/video-replay/video-replay.service';
import { DifferencesHashMap } from '@common/classes/differences-hashmap';
import { GameConstants } from '@common/classes/game-constants';
import { GameData } from '@common/classes/game-data';
import { GameRoom } from '@common/classes/game-room';
import { GameModeEvents } from '@common/enums/game-mode.gateway.variables';
import { WaitingRoomEvents } from '@common/enums/waiting-room.gateway.variables';
import { GameMode } from '@common/game-mode';
import { Subject } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class GameService {
    gameExists$ = new Subject<boolean>();
    timePosition$ = new Subject<number>();
    serverValidateResponse$ = new Subject<DifferenceTry>();
    totalDifferencesFound$ = new Subject<number>();
    userDifferencesFound$ = new Subject<number>();
    timer$ = new Subject<number>();
    gameFinished$ = new Subject<EndGame>();
    gameRoom$ = new Subject<GameRoom>();
    gameDeleted$ = new Subject<boolean>();
    abandoned$ = new Subject<string>();
    hint$ = new Subject<{ imageData: string; left: boolean }>();
    cheatModeResponse$ = new Subject<boolean>();
    gameRoom: GameRoom;
    differencesHashMap: Map<string, DifferencesHashMap[]> = new Map<string, DifferencesHashMap[]>();
    slides: GameData[];
    gameData: GameData;
    username: string;
    gameMode: string;
    gameConstants: GameConstants;
    private canSendValidate = true;

    // eslint-disable-next-line max-params
    constructor(
        readonly socketService: CommunicationSocketService,
        private chatService: ChatService,
        private communicationService: CommunicationHttpService,
        public dialog: MatDialog,
        private videoReplayService: VideoReplayService,
    ) {
        this.getConstant();
        this.getAllGames();
    }

    getIsTyping(): boolean {
        return this.chatService.getIsTyping();
    }

    getConstant(): void {
        if (this.gameRoom) {
            this.gameConstants = this.gameRoom.gameConstants;
        }
    }

    isLimitedTimeMode(): boolean {
        return this.gameMode === GameMode.limitedTimeMode;
    }

    startGame(gameRoom: GameRoom, username: string): void {
        this.gameRoom = gameRoom;
        this.username = username;
        this.gameMode = gameRoom.gameMode;
        const slide = this.getGameData(gameRoom.userGame.gameName);
        // if (!slide) {
        //     this.dialog.open(MessageDialogComponent, {
        //         panelClass: 'custom-modal',
        //         data: { message: 'Jeu introuvable' },
        //     });
        //     return;
        // }
        if (slide) {
            this.gameData = slide;
        }
        // eslint-disable-next-line @typescript-eslint/no-magic-numbers
        if (this.gameRoom.userGame.chosenDifference !== -1) {
            this.gameData.differenceMatrix = this.gameData.differenceHashMap[this.gameRoom.userGame.chosenDifference].differenceMatrix;
        }
        this.gameRoom$.next(this.gameRoom);
        if (this.gameRoom.userGame.creator === this.username) {
            this.socketService.send(WaitingRoomEvents.Start, this.gameRoom.roomId);
        }
        if (this.gameMode === GameMode.limitedTimeMode) {
            this.slides = this.slides.filter((game) => game.name !== this.gameData.name);
            this.gameDeletedSocket();
        }
        this.chatService.handleMessage();
        this.handleSocket();
    }

    observeGame(gameRoom: GameRoom, username: string) {
        this.gameRoom = gameRoom;
        this.username = username;
        const slide = this.getGameData(gameRoom.userGame.gameName);
        // if (!slide) {
        //     this.dialog.open(MessageDialogComponent, {
        //         panelClass: 'custom-modal',
        //         data: { message: 'Jeu introuvable' },
        //     });
        //     return;
        // }
        if (slide) {
            this.gameData = slide;
        }
        // eslint-disable-next-line @typescript-eslint/no-magic-numbers
        if (this.gameRoom.userGame.chosenDifference !== -1) {
            this.gameData.differenceMatrix = this.gameData.differenceHashMap[this.gameRoom.userGame.chosenDifference].differenceMatrix;
        }
        this.gameMode = gameRoom.gameMode;
        this.gameRoom$.next(this.gameRoom);
        if (this.gameMode === GameMode.limitedTimeMode) {
            this.getAllGames();
            this.gameDeletedSocket();
        }
        this.chatService.handleMessage();
        this.handleSocket();
        this.socketService.send(WaitingRoomEvents.ObservingGame, this.gameRoom);
    }

    gameDeletedSocket(): void {
        this.socketService.on(WaitingRoomEvents.GameCanceled, (gameName: string) => {
            this.slides = this.slides.filter((game) => game.name !== gameName);
        });
    }

    sendMessage(message: string, username: string): void {
        this.chatService.sendMessage(message, username, this.gameRoom.roomId);
    }

    sendServerDifference(differenceMatrix: number[][], username: string) {
        if (this.username === username) {
            this.socketService.send(GameModeEvents.SendDifference, { differenceMatrix, roomId: this.gameRoom.roomId });
        }
    }

    turnOffWaitingSocket(): void {
        this.socketService.off(WaitingRoomEvents.GameInfo);
        this.socketService.off(WaitingRoomEvents.GameCreated);
        this.socketService.off(WaitingRoomEvents.PlayerAccepted);
        this.socketService.off(WaitingRoomEvents.PlayerRejected);
        this.socketService.off(WaitingRoomEvents.GameCanceled);
    }

    turnOffGameSocket(): void {
        this.socketService.off(GameModeEvents.DifferenceValidated);
        this.socketService.off(GameModeEvents.GameFinished);
        this.socketService.off(GameModeEvents.Abandoned);
        this.socketService.off(GameModeEvents.Timer);
        this.socketService.off(GameModeEvents.CheatMode);
        this.socketService.off(GameModeEvents.NextGame);
    }

    getAllGames() {
        this.communicationService.getAllGames().subscribe((games) => {
            games.forEach((game) => {
                this.communicationService.getGame(game.name).subscribe((gameData) => {
                    game.differenceMatrix = gameData.differenceMatrix;
                    game.differenceHashMap = gameData.differenceHashMap;
                });
            });
            this.slides = games;
        });
    }

    abortGame(): void {
        if (this.socketService.isSocketAlive() && this.gameRoom.userGame.creator === this.username) {
            this.socketService.send(WaitingRoomEvents.AbortGameCreation, this.gameRoom.roomId);
        } else if (this.socketService.isSocketAlive() && this.gameRoom) {
            this.socketService.send(WaitingRoomEvents.LeaveGame, { roomId: this.gameRoom.roomId, username: this.username });
        }
    }

    sendServerValidate(differencePos: Vec2): void {
        if (!this.canSendValidate) {
            return;
        }
        this.socketService.send(GameModeEvents.ValidateDifference, {
            differencePos,
            roomId: this.gameRoom.roomId,
            username: this.username,
            validated: true,
        });
        this.canSendValidate = false;
    }

    findDifference(pos: Vec2): number[][] {
        for (const differenceHashMap of this.gameData.differenceHashMap) {
            // eslint-disable-next-line @typescript-eslint/no-magic-numbers
            if (differenceHashMap.differenceMatrix[pos.y][pos.x] !== -1) {
                return differenceHashMap.differenceMatrix;
            }
        }
        return this.gameData.differenceHashMap[0].differenceMatrix;
    }

    changeTime(number: number): void {
        if (this.socketService.isSocketAlive()) {
            this.socketService.send(GameModeEvents.ChangeTime, { roomId: this.gameRoom.roomId, time: number });
        }
    }

    reset(): void {
        this.gameRoom = undefined as unknown as GameRoom;
        this.canSendValidate = true;
        this.username = '';
        this.totalDifferencesFound$ = new Subject<number>();
        this.userDifferencesFound$ = new Subject<number>();
        this.timer$ = new Subject<number>();
        this.gameFinished$ = new Subject<EndGame>();
        this.gameRoom$ = new Subject<GameRoom>();
        this.serverValidateResponse$ = new Subject<DifferenceTry>();
        this.abandoned$ = new Subject<string>();
        this.socketService.off(WaitingRoomEvents.Started);
        this.socketService.off(GameModeEvents.DifferenceValidated);
        this.socketService.off(GameModeEvents.GameFinished);
        this.socketService.off(GameModeEvents.Abandoned);
        this.socketService.off(GameModeEvents.Timer);
        this.socketService.off(GameModeEvents.CheatMode);
        this.socketService.off(GameModeEvents.Hint);
    }

    abandonGame(): void {
        if (this.socketService.isSocketAlive() && this.gameRoom.userGame.currentPlayers?.some((player) => player.username === this.username)) {
            this.socketService.send(GameModeEvents.Abandoned, { roomId: this.gameRoom.roomId, username: this.username });
        } else if (this.socketService.isSocketAlive()) {
            this.socketService.send(GameModeEvents.ObserverLeaveGame, { roomId: this.gameRoom.roomId, username: this.username });
        }
    }

    loadNextGame(): void {
        if (this.slides.length !== 0 && this.gameRoom.userGame.creator === this.username) {
            const game = this.slides[this.generateRandomNumber(this.slides.length)];
            this.gameRoom.userGame.gameName = game.name;
            this.gameRoom.userGame.chosenDifference = this.generateRandomNumber(game.nbDifference);
            this.socketService.send(GameModeEvents.NextGame, this.gameRoom);
        } else if (this.slides.length === 0) {
            this.socketService.send(GameModeEvents.EndGame, this.gameRoom.roomId);
        }
    }

    sendHint(imageData: string, receiver: string, left: boolean): void {
        if (this.socketService.isSocketAlive()) {
            this.socketService.send(GameModeEvents.Hint, { imageData, sender: this.username, receiver, roomId: this.gameRoom.roomId, left });
        }
    }

    isCheatModeOn(): void {
        if (!this.gameRoom.gameConstants.cheatMode) {
            this.cheatModeResponse$.next(false);
        }
        this.socketService.send(GameModeEvents.CheatMode, this.gameRoom.roomId);
    }

    // private updateBestTime(gameFinished: boolean, winner: boolean): void {
    //     this.configHttpService.getBestTime(this.gameRoom.userGame.gameData.name).subscribe((bestTimes) => {
    //         if (!bestTimes) return;
    // eslint-disable-next-line max-len
    //         const actualBestTime = this.gameRoom.userGame.currentPlayers.length > 1 ? bestTimes.vsBestTimes[2].time : bestTimes.soloBestTimes[2].time;
    //         if (this.gameRoom.userGame.timer < actualBestTime && winner && gameFinished && !this.isAbandoned) {
    //             const newBestTime = new NewBestTime();
    //             newBestTime.gameName = this.gameRoom.userGame.gameData.name;
    //             newBestTime.time = this.gameRoom.userGame.timer;
    //             newBestTime.name = this.username;
    //             newBestTime.isSolo = this.gameRoom.userGame.currentPlayers.length === 1;
    //             this.configHttpService.updateBestTime(this.gameRoom.userGame.gameData.name, newBestTime).subscribe((position) => {
    //                 if (position === NOT_TOP3) return;
    //                 this.timePosition$.next(position);
    //             });
    //         }
    //     });
    // }

    private getGameData(gameName: string): GameData | undefined {
        return this.slides.find((game) => game.name === gameName);
    }

    private handleSocket(): void {
        this.socketService.on(WaitingRoomEvents.Started, () => {
            this.gameRoom$.next(this.gameRoom);
            this.videoReplayService.setGameData(this.gameData);
            this.videoReplayService.recordEvent({
                action: 'gameStart',
                imageData1: '',
                imageData2: '',
                timestamp: Date.now(),
            });
        });

        this.socketService.on(GameModeEvents.NextGame, (gameRoom: GameRoom) => {
            this.gameRoom = gameRoom;
            const slideIndex = this.slides.findIndex((game) => game.name === gameRoom.userGame.gameName);
            const slide = this.slides.splice(slideIndex, 1)[0];
            if (!slide) {
                this.dialog.open(MessageDialogComponent, {
                    panelClass: 'custom-modal',
                    data: { message: 'Problème interne du serveur, svp réessayer ' },
                });
                return;
            }
            this.gameData = slide;
            this.gameData.differenceMatrix = this.gameData.differenceHashMap[this.gameRoom.userGame.chosenDifference].differenceMatrix;
            this.gameRoom$.next(this.gameRoom);
        });

        this.socketService.on(GameModeEvents.Hint, (data: { imageData: string; sender: string; receiver: string; left: boolean }) => {
            if (
                data.receiver === this.username ||
                (data.receiver === null && !this.gameRoom.userGame.observers?.some((observer) => observer.username === this.username)) ||
                data.sender === this.username
            ) {
                this.hint$.next({ imageData: data.imageData, left: data.left });
            }
        });

        this.socketService.on(GameModeEvents.DifferenceValidated, (differenceTry: DifferenceTry) => {
            if (differenceTry.validated) {
                this.gameRoom.userGame.nbDifferenceFound++;
                this.totalDifferencesFound$.next(this.gameRoom.userGame.nbDifferenceFound);
                this.gameRoom.userGame.differenceFoundByPlayers = differenceTry.everyoneScore;
            }
            this.serverValidateResponse$.next(differenceTry);
        });

        this.socketService.on(GameModeEvents.CheatMode, (isCheatModeOn: boolean) => {
            this.cheatModeResponse$.next(isCheatModeOn);
        });

        this.socketService.on(GameModeEvents.GameFinished, (endGame: EndGame) => {
            this.gameFinished$.next(endGame);
            this.videoReplayService.recordEvent({
                action: 'gameEnd',
                imageData1: '',
                imageData2: '',
                timestamp: Date.now(),
            });
        });

        this.socketService.on(WaitingRoomEvents.ObservingGame, (gameRoom: GameRoom) => {
            this.gameRoom.userGame.observers = gameRoom.userGame.observers;
            this.gameRoom$.next(this.gameRoom);
        });

        this.socketService.on(GameModeEvents.Abandoned, (data: { gameRoom: GameRoom; username: string }) => {
            this.gameRoom = data.gameRoom;
            this.gameRoom$.next(this.gameRoom);
            this.abandoned$.next(data.username);
        });

        this.socketService.on(GameModeEvents.Timer, (timer: number) => {
            this.gameRoom.userGame.timer = timer;
            this.timer$.next(timer);
            this.canSendValidate = true;
        });
    }

    // private limitedTimeGameAbandoned(gameRoom: GameRoom): void {
    //     this.gameRoom = gameRoom;
    //     this.gameRoom$.next(this.gameRoom);
    // }

    private generateRandomNumber(nbOfDifferences: number): number {
        return Math.floor(Math.random() * nbOfDifferences);
    }
}
